api/admin/controller/library_settings.py

```diff
@@ -216,29 +221,83 @@ def get_extra_geographic_information(self, value):
+            validated_value = self._validate_setting(library, setting, validator)
+
+            if isinstance(validated_value, ProblemDetail):
+                # Validation failed -- return a ProblemDetail.
+                return validated_value
+
+            # Validation succeeded -- set the new value.
+            ConfigurationSetting.for_library(setting['key'], library).value = validated_value
+
+    def _validate_setting(self, library, setting, validator=None):
+        """Validate the incoming value for a single library setting.
+
+        :param library: A Library
+        :param setting: Configuration data for one of the library's settings.
+        :param validator: A validation object for data of this type.
+        """
+        # TODO: there are some opportunities for improvement here:
+        # * There's no standard interface for validators.
+        # * We can handle settings that are lists of certain types (language codes,
+        #   geographic areas), but not settings that are a single value of that type
+        #   (_one_ language code or geographic area). Sometimes there's even an implication
+        #   that a certain data type ('geographic') _must_ mean a list.
+        # * A list value is returned as a JSON-encoded string. It
+        #   would be better to keep that as a list for longer in case
+        #   controller code needs to look at it.
+        format = setting.get('format')
+        type = setting.get('type')
+
+        # In some cases, if there is no incoming value we can use a
+        # default value or the current value.
+        #
+        # When the configuration item is a list, we can't do this
+        # because an empty list may be a valid value.
+        current_value = self.current_value(setting, library)
+        default_value = setting.get('default') or current_value
+
+        if format == "geographic":
+            value = self.list_setting(setting)
+            value = validator.validate_geographic_areas(value, self._db)
+        elif format == "announcements":
+            value = self.list_setting(setting)
+            value = validator.validate(value)
+        elif type == "list":
+            value = self.list_setting(setting)
+            if format == "language-code":
+                value = json.dumps([LanguageCodes.string_to_alpha_3(language) for language in json.loads(value)])
+        else:
+            if type == "image":
+                value = self.image_setting(setting) or default_value
             else:
-                default = setting.get('default')
-                value = flask.request.form.get(setting['key'], default)
-
-            ConfigurationSetting.for_library(setting['key'], library).value = value
+                value = self.scalar_setting(setting) or default_value
+        return value
 
-    def current_value(self, setting, library):
-        return ConfigurationSetting.for_library(setting['key'], library).value
+    def scalar_setting(self, setting):
```

Comment:
Why "scalar"?

Reply:
It's one of those technical terms where I haven't been able to find a less fancy term  that's equally precise. A scalar value means a single value, as opposed to a more complicated grouping of values such as a list. So this is in opposition to `list_setting`.